sbol_utilities/sbol3_genbank_conversion.py

```diff
@@ -1,48 +1,112 @@
+        """Reads 2 CSV Files containing mappings for converting between GenBank and SO ontologies roles
+        :param gb2so_csv: path to read genbank to so conversion csv file
+        :param so2gb_csv: path to read so to genbank conversion csv file
+        """
+        with open(gb2so_csv, mode='r') as csv_file:
+            csv_reader = csv.DictReader(csv_file)
+            for row in csv_reader:
+                GB2SO_MAPPINGS[row["GenBank_Ontology"]] = row["SO_Ontology"]
+        with open(so2gb_csv, mode='r') as csv_file:
+            csv_reader = csv.DictReader(csv_file)
+            for row in csv_reader:
+                SO2GB_MAPPINGS[row["SO_Ontology"]] = row["GenBank_Ontology"]
+
+
+    def convert_genbank_to_sbol3(self, gb_file: str, sbol3_file: str = "sbol3.out", namespace: str =
+                                TEST_NAMESPACE, write: bool = False) -> sbol3.Document:
+        """Convert a GenBank document on disk into an SBOL3 document
+        Specifically, the GenBank document is first imported to SBOL2, then converted from SBOL2 to SBOL3
+
+        :param gb_file: path to read GenBank file from
+        :param sbol3_file: path to write SBOL3 file to, if write set to true
+        :param namespace: URIs of Components will be set to {namespace}/{genbank_id},
+                          defaults to "https://test.sbol3.genbank/" 
+        :param write: writes the generated sbol3 document in SORTED_NTRIPLES
+                          format to provided sbol3_file path
+        :return: SBOL3 document containing converted materials
+        """
+        # create sbol3 document, and record parser handler for gb file
+        sbol3.set_namespace(namespace)
+        doc = sbol3.Document()
+        # access records by parsing gb file using SeqIO class
+        records = list(SeqIO.parse(gb_file, "genbank").records)
+        for record in records:
+            # NOTE: Currently we assume only linear or circular topology is possible
+            COMP_TYPES.append(sbol3.SO_LINEAR 
+                              if record.annotations['topology'] == "linear" else sbol3.SO_CIRCULAR)
+            comp = sbol3.Component(identity=record.name,
+                                   types=COMP_TYPES,
+                                   roles=COMP_ROLES,
+                                   description=record.description)
+            doc.add(comp)
+            # NOTE: Currently we use a fixed method of encoding (IUPAC)
+            seq = sbol3.Sequence(identity=record.name + "_sequence",
+                                 elements=str(record.seq.lower()),
+                                 encoding=SEQUENCE_ENCODING)
+            doc.add(seq)
+            comp.sequences = [seq]
+            if record.features:
+                # create updated py dict to store mappings between gb and so ontologies
+                self.create_GB2SO_role_mappings(gb2so_csv=GB2SO_MAPPINGS_CSV, so2gb_csv=SO2GB_MAPPINGS_CSV)
```

Comment:
does this mean that both CSV files are read in for each record that gets processed? Is it necessary to read these files for every record processed?

Reply:
Yes, I agree it wouldn't be efficient to parse both CSVs for every record.
* I have fixed this to parse only once for a GenBank file.
* Also added a flag parameter to only let 1 of the 2 CSVs be parsed, based on our direction of conversion.